Penambahan fitur bagi pengguna terdaftar

noaione opened the pull request:
Relevan: #8 (fix #8)

### Contoh
`KBBI('roh', email="REDACTED", password="REDACTED")`
```py
roh
bentuk tidak baku: ruh
Etimologi: [Arab] <n> <sg> <f/m> » رُوْحٌ (rūh): tiupan; sesuatu yang membuat manusia dapat hidup
1. (n)  sesuatu (unsur) yang ada dalam jasad yang diciptakan Tuhan sebagai penyebab adanya hidup (kehidupan); nyawa: jika -- sudah berpisah dari badan, berakhirlah kehidupan seseorang
2. (n)  makhluk hidup yang tidak berjasad, tetapi berpikiran dan berperasaan (malaikat, jin, setan, dan sebagainya)
3. (n) (ki)  semangat; spirit: kedamaian bagi seluruh warga sesuai dengan -- Islam
Gabungan Kata: roh Kudus; roh suci
```

`KBBI('coba', email="REDACTED", password="REDACTED")`
```py
co.ba
1. (v)  silakan; sudilah; tolong (untuk menghaluskan suruhan atau ajakan): -- lihat kalau-kalau ia ada di rumah
2. (p)  kata afektif untuk menyatakan perasaan jengkel dan sebagainya: --, orang apa ini, begitu saja tidak bisa
3. (p)  seandainya; jika: -- tidak kutolong, celaka engkau
Kata Turunan: coba-coba; cobaan; mencoba; mencoba-coba; mencobai; mencobakan; pencoba; pencobaan; percobaan
Gabungan Kata: coba dan ralat
Peribahasa (mengandung [co.ba]): coba-coba menanam mumbang, kalau tumbuh sunting negeri, dicoba-coba bertanam mumbang, moga-moga tumbuh kelapa
```

Mungkin nanti bisa diganti beberapa variabelnya

Commit:
Penambahan fitur bagi pengguna terdaftar
- etimologi
- kata turunan
- gabungan kata
- peribahasa

Commit:
buat email/password menjadi opsional

Commit:
Linting dan hapus UA Chrome

laymonage commented:
Wah, padahal saya baru saja mau rilis versi 0.4.0 :sweat_smile:. Terima kasih! Saya coba tinjau dulu, ya.

laymonage commented on kbbi/kbbi.py:
```diff
@@ -22,15 +23,23 @@ class KBBI:
     host = "https://kbbi.kemdikbud.go.id"
 
-    def __init__(self, kueri):
+    def __init__(self, kueri, email=None, password=None):
```
Kalau seperti ini, sayangnya proses autentikasi akan dilakukan setiap kali ingin mengambil laman KBBI.

Saya rasa urusan autentikasi sebaiknya dilakukan dengan objek dari *class* terpisah. Mungkin kita bisa membuat *class* `AutentikasiKBBI` yang menangani urusan autentikasi (pindahkan metode `_autentikasi()` ke sana). Lalu, kelas `KBBI` akan menerima objek `AutentikasiKBBI` tersebut (yang di dalamnya terdapat `Session`). Dengan begitu, pengguna cukup buat objek `AutentikasiKBBI` **satu kali** (sehingga `_autentikasi()` hanya dipanggil satu kali), dan menggunakan objek tersebut untuk pembuatan objek-objek `KBBI` berikutnya.

laymonage replied:
Oh ya, kalau hanya begini sayangnya pemanggilan via CLI tetap harus melakukan autentikasi berkali-kali. Mungkin kita bisa tambahkan opsi autentikasi dengan `.AspNet.ApplicationCookie` agar pemanggilan via CLI tidak harus *login* terus menerus.

noaione replied:
Iya, untuk sementara harus call `_autentikasi()` tiap summon class `KBBI`.
Mungkin bisa coba implemen local save untuk cookies di `localappdata` atau semacamnya untuk yang CLI.

laymonage commented on kbbi/kbbi.py:
```diff
@@ -54,6 +63,33 @@ def _cek_galat(self, laman):
         if "Entri tidak ditemukan." in laman.text:
             raise TidakDitemukan(self.nama)
 
+    def _autentikasi(self, email, password):
```
Secara umum, metode ini sudah oke. Sebenarnya bisa pakai `BeautifulSoup` supaya tidak perlu keajaiban regex, tetapi berhubung kita tidak perlu mengolah HTML halaman login, begini saja lebih baik.

laymonage commented on kbbi/kbbi.py:
```diff
@@ -145,11 +207,77 @@ def _init_varian(self, judul):
+        kata_turunan = None
+        gabungan_kata = None
+        peribahasa = None
+        for le in lain_lain:
```
Saya melihat pola untuk ketiga hal ini... mungkin bisa disederhanakan? :wink:

laymonage replied:
```suggestion

terkait = {
    "Kata Turunan": self.kata_turunan,
    "Gabungan Kata": self.gabungan_kata,
    "Peribahasa": self.peribahasa,
    "Kiasan": self.kiasan,
}

for le in lain_lain:
    if not le:
        continue
    le_txt = le.text.strip()
    for jenis, daftar in terkait.items():
        if jenis in le_txt:
            kumpulan = le.next_sibling
            if kumpulan:
                 kumpulan = kumpulan.find_all("a")
                 daftar.extend([k.text for k in kumpulan if k])
```

laymonage commented on kbbi/kbbi.py:
```diff
@@ -365,6 +609,20 @@ def _parse_args(args):
+    parser.add_argument(
+        "-U",
+        "--username",
+        help="gunakan indentasi sebanyak N untuk serialisasi JSON",
```
Ups, pesan `help`-nya belum diganti :smile:

laymonage commented on kbbi/kbbi.py:
```diff
@@ -381,7 +639,7 @@ def main(argv=None):
     try:
-        laman = KBBI(args.laman)
+        laman = KBBI(args.laman, email=args.username, password=args.password)
     except (TidakDitemukan, TerjadiKesalahan, BatasSehari) as e:
```
`GagalAutentikasi` juga akan perlu dimasukkan ke dalam sini. Mungkin ini saatnya kita buat abstraksi untuk galat yang kita buat, misal `class Galat(Exception)` kemudian semua *class* galat yang dibuat akan diturunkan dari `Galat`.

laymonage reviewed (changes requested):
Kerja bagus, keren sekali! Saya belum sempat mencobanya sih, tetapi melihat kodenya sekilas, sepertinya sudah cukup baik. Ini hasil tinjauan pertama saya.

Commit:
Buat implementasi "penyimpanan cookie"

Commit:
Tambah kata kiasan (https://kbbi.kemdikbud.go.id/entri/makan)

Commit:
Buat objek AutentikasiKBBI untuk menggunakan sesi login berkali-kali

Commit:
Reformat dan Linting

laymonage commented on kbbi/kbbi.py:
```diff
@@ -23,26 +24,45 @@ class KBBI:
 
+    def __ambil_cookies(self):
+        env = {"nt": "LOCALAPPDATA"}
+        save_folder = os.path.join(
```
Ketimbang pakai `os.path`, mungkin lebih baik pakai `pathlib`. Direktori `home` bisa didapat dengan [`pathlib.Path.home`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.home).

laymonage commented on kbbi/kbbi.py:
```diff
@@ -23,26 +24,45 @@ class KBBI:
+            return
+        if os.path.isfile(f"{save_folder}/cookie.txt"):
+            with open(f"{save_folder}/cookie.txt", "r") as fp:
+                self.sesi.headers.update({"Cookie": fp.read()})
```
Sepertinya cara yang baik untuk mengatur *cookie* pada objek `Session` adalah dengan memanfaatkan objek `RequestsCookieJar` (dapat diakses dengan `self.sesi.cookies`) dan menggunakan *method* [`set`](https://2.python-requests.org/en/v2.8.1/api/#requests.cookies.RequestsCookieJar.set) (atau `update`/semacamnya). 

Ketimbang `.txt` dengan format `key=value`, mungkin lebih baik pakai JSON dan memanfaatkan *method* [`dict_from_cookiejar`](https://2.python-requests.org/en/v2.8.1/api/#requests.utils.dict_from_cookiejar) serta `cookiejar_from_dict`. Kemudian, `dict` dimuat/disimpan dengan [`json.load`](https://docs.python.org/3/library/json.html#json.load) dan `json.dump`. Silakan saja ingin menyimpan seluruh *cookie* atau `.AspNet.ApplicationCookie` saja.

Kalau dirasa terlalu merepotkan, saya bisa bantu melanjutkan.

noaione replied:
Mungkin bisa dilanjutkan, saya ubah dulu `os.path` ke [`pathlib.Path`](https://docs.python.org/3/library/pathlib.html) dulu aja

laymonage replied:
Baik, saya lanjutkan ya.

laymonage commented on kbbi/kbbi.py:
```diff
@@ -639,11 +705,24 @@ def main(argv=None):
+            auth = AutentikasiKBBI(args.username, args.password)
+        laman = KBBI(args.laman, auth)
+    except (
+        TidakDitemukan,
```
Buat *class* berikut:

```python
class Galat(Exception):
    pass
```

Kemudian buat agar `TidakDitemukan`, `TerjadiKesalahan`, `BatasSehari`, dan `GagalAutentikasi` menjadi *subclass* dari `Galat`. Lalu, di sini cukup:

```python
except Galat as e:
    print(e)
```

laymonage commented on kbbi/kbbi.py:
```diff
@@ -23,26 +24,45 @@ class KBBI:
+        if not os.path.isdir(save_folder):
+            return
+        aspcookie = self.sesi.cookies.get(".AspNet.ApplicationCookie")
+        if aspcookie:
```
Saya penasaran, kalau *cookie*-nya sudah kedaluwarsa, apa yang akan terjadi? :thinking:

laymonage replied:
Oh ya, saya lupa tambahkan. Mungkin akan lebih baik juga kalau ada opsi CLI untuk menghapus/mengabaikan *cookie* yang sudah disimpan.

laymonage reviewed (changes requested):
Terima kasih sudah merevisi PR ini. Berikut hasil tinjauan kedua saya.

Commit:
Sesuaikan dengan saran
- Gunakan `pathlib.Path` daripada `os.path`
- Buat class utama untuk Galat
- Ubah bagian "lain-lain"

laymonage committed:
Sederhanakan kode init makna

laymonage committed:
Sederhanakan kode init etimologi

laymonage committed:
Pecah pemrosesan str untuk entri terkait dalam Entri

laymonage committed:
Sederhanakan proses pengecekan terautentikasi

laymonage committed:
Sederhanakan kode pemeriksaan varian

laymonage committed:
Kurangi indentasi dengan return terlebih dahulu

laymonage committed:
Ubah lain_lain menjadi terkait dan sederhanakan kode

laymonage committed:
Gabung variabel untuk entri-entri terkait menjadi satu dict

laymonage committed:
Sederhanakan pemrosesan etimologi

laymonage committed:
Pecah arti etimologi menjadi list

laymonage committed:
Gunakan asal kata untuk representasi etimologi

laymonage committed:
Buat pengolahan etimologi menjadi lebih aman

Antisipasi jika ada atribut yang tidak lengkap

Format juga sedikit diubah agar lebih sesuai dengan situs asli

laymonage committed:
Perbaiki pengolahan prakategorial jika terautentikasi

laymonage committed:
Perbaiki kasus uji untuk entri yang tidak bermakna

laymonage committed:
Ubah cookies -> kuki

laymonage committed:
Pindahkan method ambil cookie ke AutentikasiKBBI

laymonage committed:
Munculkan galat jika autentikasi gagal

laymonage committed:
Pisahkan autentikasi dari program utama

laymonage committed:
Gunakan appdirs untuk mengolah path kuki

laymonage committed:
Gunakan JSON untuk menyimpan kuki

laymonage committed:
Perbaiki penanganan galat ketika kuki belum ada

laymonage committed:
Tambahkan opsi untuk menghapus kuki

laymonage committed:
Tambahkan opsi untuk menonaktifkan fitur khusus pengguna

laymonage committed:
Tambahkan parameter untuk mengatur lokasi kuki

laymonage committed:
Buat metode pada AutentikasiKBBI menjadi publik

laymonage committed:
Tambahkan opsi untuk tidak menampilkan entri terkait

laymonage committed:
Sederhanakan docstring

laymonage committed:
Refaktor kode

laymonage committed:
Masukkan variabel lokasi kuki ke dalam AutentikasiKBBI

laymonage committed:
Dokumentasikan penggunaan fitur khusus pengguna di README

laymonage committed:
Perbaiki pesan bantuan

laymonage committed:
Gunakan flag untuk menonaktifkan fitur pengguna

Kuki tetap digunakan dan batas pencarian tetap lebih besar

laymonage committed:
Perbaiki kasus di mana kelas pada makna bisa diklik

Hanya terjadi jika terautentikasi

laymonage committed:
Tambahkan kasus tes untuk fitur khusus pengguna

laymonage committed:
Ubah konfigurasi linting dan tes
